README
Plugin module for VLC player
============================

This plugin allows you to automatically save and restore the position
of the last played files in the VLC player.

Installation
============
  * MS Windows:
    Copy file libsrpos*_plugin.dll to {VLC installation folder}\plugins;

  * Unix/Linux and similar systems:
    Build and install plugin module:
      ./configure
      make
      make install
    Use configure options to specify VLC include/library/output paths
    (--with-vlc-*-path options).


VLC player setup
================
  1. Start VLC player;
  2. Open preferences window (menu Tools->Preferences);
  3. Select 'All' in 'Show settings';
  4. Open Control interfaces panel (Interface->Control interfaces);
  5. Check 'Save/restore position of the last played files' checkbox;
  6. Click 'Save' button;
  7. Restart VLC player.

Warning:
  This plugin is provided "as is". The author is not responsible for the
  consequences of use of this software.

Best regards,
Artem Senichev (artemsen@gmail.com)
               http://code.google.com/p/vlc-srpos-plugin


ChangeLog
libsrpos_plugin 0.3 2013-10-04
------------------------------
 * support to rememeber up to 100 last files position
 * VLC 2.1 (Windows build)

libsrpos_plugin 0.2 2012-11-15
------------------------------
 * support to rememeber up to 10 last files position

libsrpos_plugin 0.1 2012-11-10
------------------------------
 * first official release! :-)


vlc_interface.h
/*****************************************************************************
 * vlc_interface.h: interface access for other threads
 * This library provides basic functions for threads to interact with user
 * interface, such as message output.
 *****************************************************************************
 * Copyright (C) 1999, 2000 VLC authors and VideoLAN
 * $Id: 9bc6241b7eb84459a28f3fee32bff129d74bad80 $
 *
 * Authors: Vincent Seguin <seguin@via.ecp.fr>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

#ifndef VLC_INTF_H_
#define VLC_INTF_H_

# ifdef __cplusplus
extern "C" {
# endif

typedef struct intf_dialog_args_t intf_dialog_args_t;

/**
 * \file
 * This file contains structures and function prototypes for
 * interface management in vlc
 */

/**
 * \defgroup vlc_interface Interface
 * These functions and structures are for interface management
 * @{
 */

typedef struct intf_sys_t intf_sys_t;

/** Describe all interface-specific data of the interface thread */
typedef struct intf_thread_t
{
    VLC_COMMON_MEMBERS

    struct intf_thread_t *p_next; /** LibVLC interfaces book keeping */

    /* Specific interfaces */
    intf_sys_t *        p_sys;                          /** system interface */

    /** Interface module */
    module_t *   p_module;

    /** Specific for dialogs providers */
    void ( *pf_show_dialog ) ( struct intf_thread_t *, int, int,
                               intf_dialog_args_t * );

    config_chain_t *p_cfg;
} intf_thread_t;

/** \brief Arguments passed to a dialogs provider
 *  This describes the arguments passed to the dialogs provider. They are
 *  mainly used with INTF_DIALOG_FILE_GENERIC.
 */
struct intf_dialog_args_t
{
    intf_thread_t *p_intf;
    char *psz_title;

    char **psz_results;
    int  i_results;

    void (*pf_callback) ( intf_dialog_args_t * );
    void *p_arg;

    /* Specifically for INTF_DIALOG_FILE_GENERIC */
    char *psz_extensions;
    bool b_save;
    bool b_multiple;

    /* Specific to INTF_DIALOG_INTERACTION */
    struct interaction_dialog_t *p_dialog;
};

VLC_API int intf_Create( vlc_object_t *, const char * );
#define intf_Create(a,b) intf_Create(VLC_OBJECT(a),b)

VLC_API void libvlc_Quit( libvlc_int_t * );

/**
 * \defgroup vlc_subscription Log messages subscription
 * These functions deal with log messages.
 * @{
 */

/**
 * Message logging callback signature.
 * \param data data pointer as provided to vlc_msg_SetCallback().
 * \param type message type (VLC_MSG_* values from enum vlc_log_type)
 * \param item meta informations
 * \param fmt format string
 * \param args format string arguments
 *
  typedef void (*vlc_log_cb) (void *data, int type, const *item,
                            const char *fmt, va_list args);
  */
//VLC_API void vlc_LogSet(libvlc_int_t *, vlc_log_cb cb, void *data);
  
  //typedef struct msg_subscription { } msg_subscription_t;
#define vlc_Subscribe(sub,cb,data) ((sub), (cb), (data))
#define vlc_Unsubscribe(sub) ((void)(sub))
  
/*@}*/

#if defined( _WIN32 ) && !VLC_WINSTORE_APP
#    define CONSOLE_INTRO_MSG \
         if( !getenv( "PWD" ) ) /* detect Cygwin shell or Wine */ \
         { \
         AllocConsole(); \
         freopen( "CONOUT$", "w", stdout ); \
         freopen( "CONOUT$", "w", stderr ); \
         freopen( "CONIN$", "r", stdin ); \
         } \
         msg_Info( p_intf, "VLC media player - %s", VERSION_MESSAGE ); \
         msg_Info( p_intf, "%s", COPYRIGHT_MESSAGE ); \
         msg_Info( p_intf, _("\nWarning: if you cannot access the GUI " \
                             "anymore, open a command-line window, go to the " \
                             "directory where you installed VLC and run " \
                             "\"vlc -I qt\"\n") )
#else
#    define CONSOLE_INTRO_MSG (void)0
#endif

/* Interface dialog ids for dialog providers */
typedef enum vlc_dialog {
    INTF_DIALOG_FILE_SIMPLE = 1,
    INTF_DIALOG_FILE,
    INTF_DIALOG_DISC,
    INTF_DIALOG_NET,
    INTF_DIALOG_CAPTURE,
    INTF_DIALOG_SAT,
    INTF_DIALOG_DIRECTORY,

    INTF_DIALOG_STREAMWIZARD,
    INTF_DIALOG_WIZARD,

    INTF_DIALOG_PLAYLIST,
    INTF_DIALOG_MESSAGES,
    INTF_DIALOG_FILEINFO,
    INTF_DIALOG_PREFS,
    INTF_DIALOG_BOOKMARKS,
    INTF_DIALOG_EXTENDED,

    INTF_DIALOG_POPUPMENU = 20,
    INTF_DIALOG_AUDIOPOPUPMENU,
    INTF_DIALOG_VIDEOPOPUPMENU,
    INTF_DIALOG_MISCPOPUPMENU,

    INTF_DIALOG_FILE_GENERIC = 30,
    INTF_DIALOG_INTERACTION = 50,

    INTF_DIALOG_UPDATEVLC = 90,
    INTF_DIALOG_VLM,

    INTF_DIALOG_EXIT = 99
} vlc_dialog_t;

/* Useful text messages shared by interfaces */
#define INTF_ABOUT_MSG LICENSE_MSG

#define EXTENSIONS_AUDIO_CSV "3ga", "669", "a52", "aac", "ac3", "adt", "adts", "aif", "aifc", "aiff", \
                         "amr", "aob", "ape", "awb", "caf", "dts", "flac", "it", "kar", \
                         "m4a", "m4p", "m5p", "mka", "mlp", "mod", "mpa", "mp1", "mp2", "mp3", "mpc", "mpga", \
                         "oga", "ogg", "oma", "opus", "qcp", "ra", "rmi", "s3m", "spx", "thd", "tta", \
                         "voc", "vqf", "w64", "wav", "wma", "wv", "xa", "xm"

#define EXTENSIONS_VIDEO_CSV "3g2", "3gp", "3gp2", "3gpp", "amv", "asf", "avi", "divx", "drc", "dv", \
                             "f4v", "flv", "gvi", "gxf", "iso", \
                             "m1v", "m2v", "m2t", "m2ts", "m4v", "mkv", "mov",\
                             "mp2", "mp2v", "mp4", "mp4v", "mpe", "mpeg", "mpeg1", \
                             "mpeg2", "mpeg4", "mpg", "mpv2", "mts", "mtv", "mxf", "mxg", "nsv", "nuv", \
                             "ogg", "ogm", "ogv", "ogx", "ps", \
                             "rec", "rm", "rmvb", "tod", "ts", "tts", "vob", "vro", \
                             "webm", "wm", "wmv", "wtv", "xesc"

#define EXTENSIONS_AUDIO \
    "*.3ga;" \
    "*.669;" \
    "*.a52;" \
    "*.aac;" \
    "*.ac3;" \
    "*.adt;" \
    "*.adts;" \
    "*.aif;"\
    "*.aifc;"\
    "*.aiff;"\
    "*.amr;" \
    "*.aob;" \
    "*.ape;" \
    "*.awb;" \
    "*.caf;" \
    "*.dts;" \
    "*.flac;"\
    "*.it;"  \
    "*.kar;" \
    "*.m4a;" \
    "*.m4p;" \
    "*.m5p;" \
    "*.mid;" \
    "*.mka;" \
    "*.mlp;" \
    "*.mod;" \
    "*.mpa;" \
    "*.mp1;" \
    "*.mp2;" \
    "*.mp3;" \
    "*.mpc;" \
    "*.mpga;" \
    "*.oga;" \
    "*.ogg;" \
    "*.oma;" \
    "*.opus;" \
    "*.qcp;" \
    "*.ra;" \
    "*.rmi;" \
    "*.s3m;" \
    "*.spx;" \
    "*.thd;" \
    "*.tta;" \
    "*.voc;" \
    "*.vqf;" \
    "*.w64;" \
    "*.wav;" \
    "*.wma;" \
    "*.wv;"  \
    "*.xa;"  \
    "*.xm"

#define EXTENSIONS_VIDEO "*.3g2;*.3gp;*.3gp2;*.3gpp;*.amv;*.asf;*.avi;*.bin;*.divx;*.drc;*.dv;*f4v;*.flv;*.gvi;*.gvi;*.gxf;*.iso;*.m1v;*.m2v;" \
                         "*.m2t;*.m2ts;*.m4v;*.mkv;*.mov;*.mp2;*.mp2v;*.mp4;*.mp4v;*.mpe;*.mpeg;*.mpeg1;" \
                         "*.mpeg2;*.mpeg4;*.mpg;*.mpv2;*.mts;*.mtv;*.mxf;*.mxg;*.nsv;*.nuv;" \
                         "*.ogg;*.ogm;*.ogv;*.ogx;*.ps;" \
                         "*.rec;*.rm;*.rmvb;*.tod;*.ts;*.tts;*.vob;*.vro;*.webm;*.wm;*.wmv;*.wtv;*.xesc"

#define EXTENSIONS_PLAYLIST "*.asx;*.b4s;*.cue;*.ifo;*.m3u;*.m3u8;*.pls;*.ram;*.rar;*.sdp;*.vlc;*.xspf;*.wvx;*.zip;*.conf"

#define EXTENSIONS_MEDIA EXTENSIONS_VIDEO ";" EXTENSIONS_AUDIO ";" \
                          EXTENSIONS_PLAYLIST

#define EXTENSIONS_SUBTITLE "*.cdg;*.idx;*.srt;" \
                            "*.sub;*.utf;*.ass;" \
                            "*.ssa;*.aqt;" \
                            "*.jss;*.psb;" \
                            "*.rt;*.smi;*.txt;" \
                            "*.smil;*.stl;*.usf;" \
                            "*.dks;*.pjs;*.mpl2;*.mks"

/** \defgroup vlc_interaction Interaction
 * \ingroup vlc_interface
 * Interaction between user and modules
 * @{
 */

/**
 * This structure describes a piece of interaction with the user
 */
typedef struct interaction_dialog_t
{
    int             i_type;             ///< Type identifier
    char           *psz_title;          ///< Title
    char           *psz_description;    ///< Descriptor string
    char           *psz_default_button;  ///< default button title (~OK)
    char           *psz_alternate_button;///< alternate button title (~NO)
    /// other button title (optional,~Cancel)
    char           *psz_other_button;

    char           *psz_returned[1];    ///< returned responses from the user

    vlc_value_t     val;                ///< value coming from core for dialogue
    int             i_timeToGo;         ///< time (in sec) until shown progress is finished
    bool      b_cancelled;        ///< was the dialogue cancelled ?

    void *          p_private;          ///< Private interface data

    int             i_status;           ///< Dialog status;
    int             i_action;           ///< Action to perform;
    int             i_flags;            ///< Misc flags
    int             i_return;           ///< Return status

    vlc_object_t   *p_parent;           ///< The vlc object that asked
                                        //for interaction
    intf_thread_t  *p_interface;
    vlc_mutex_t    *p_lock;
} interaction_dialog_t;

/**
 * Possible flags . Dialog types
 */
#define DIALOG_GOT_ANSWER           0x01
#define DIALOG_YES_NO_CANCEL        0x02
#define DIALOG_LOGIN_PW_OK_CANCEL   0x04
#define DIALOG_PSZ_INPUT_OK_CANCEL  0x08
#define DIALOG_BLOCKING_ERROR       0x10
#define DIALOG_NONBLOCKING_ERROR    0x20
#define DIALOG_USER_PROGRESS        0x80
#define DIALOG_INTF_PROGRESS        0x100

/** Possible return codes */
enum
{
    DIALOG_OK_YES,
    DIALOG_NO,
    DIALOG_CANCELLED
};

/** Possible status  */
enum
{
    ANSWERED_DIALOG,            ///< Got "answer"
    DESTROYED_DIALOG,           ///< Interface has destroyed it
};

/** Possible actions */
enum
{
    INTERACT_NEW,
    INTERACT_UPDATE,
    INTERACT_HIDE,
    INTERACT_DESTROY
};

#define intf_UserStringInput( a, b, c, d ) (VLC_OBJECT(a),b,c,d, VLC_EGENERIC)
#define interaction_Register( t ) (t, VLC_EGENERIC)
#define interaction_Unregister( t ) (t, VLC_EGENERIC)


/** @} */
/** @} */

# ifdef __cplusplus
}
# endif
#endif


srpos.c
/*****************************************************************************
 * srpos.c : Save and restore position of the last played files
 *****************************************************************************
 * Copyright (C) 2012-2013
 * $Id:
 *
 * Authors: Artem Senichev <artemsen@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

/*****************************************************************************
 * Preamble
 *****************************************************************************/

#include <vlc_common.h>
#include <vlc_plugin.h>
#include "vlc_interface.h" /* This file not included to VLC SDK */
#include <vlc_playlist.h>

/*****************************************************************************
 * Internal state for an instance of the module
 *****************************************************************************/
typedef struct srpos_list_t srpos_list_t;
struct srpos_list_t
{
    char *psz_uri;         /* File uri */
    double f_position;     /* Last position in file */
    bool b_used;           /* Flag signaled that position has been already set */
    srpos_list_t *p_next;  /* Pointer to next item (list realisation) */
};
struct intf_sys_t
{
    srpos_list_t* p_list;  /* File position info list */
};

/*****************************************************************************
 * Forward declarations
 *****************************************************************************/
static int  Open( vlc_object_t * );
static void Close( vlc_object_t * );
static int  ItemChange( vlc_object_t *, const char *,
                        vlc_value_t, vlc_value_t, void * );
static FILE* GetSettingsFile( const bool );
static srpos_list_t* GetPositionInfo( intf_sys_t *, const char * );
static void SetPositionInfo( intf_sys_t *, const char *, double );

/*****************************************************************************
 * Module descriptor
 *****************************************************************************/
vlc_module_begin()
    set_category( CAT_INTERFACE )
    set_subcategory( SUBCAT_INTERFACE_CONTROL )
    set_shortname( MODULE_STRING )
    set_description( "Save/restore position of the last played files" )
    set_capability( "interface", 0 )
    set_callbacks( Open, Close )
vlc_module_end()

/*****************************************************************************
 * Open: initialize interface
 *****************************************************************************/
static int Open( vlc_object_t *p_this )
{
    intf_thread_t *p_intf = (intf_thread_t *)p_this;
    char *psz_sfile_data = NULL;
    char *psz_sfile_uri = NULL;
    srpos_list_t *p_prev = NULL;
    srpos_list_t *p_curr = NULL;
    FILE *p_sfile = NULL;

    /* Allocate memory */
    p_intf->p_sys = (intf_sys_t *)malloc( sizeof( intf_sys_t ) );
    if( p_intf->p_sys == NULL )
        return VLC_ENOMEM;
    p_intf->p_sys->p_list = NULL;

    /* Load files position info */
    p_sfile = GetSettingsFile( true );
    if( p_sfile )
    {
        psz_sfile_data = (char *)malloc( 1024 );
        if( psz_sfile_data )
        {
            memset( psz_sfile_data, 0, 1024 );
            while ( fgets( psz_sfile_data, 1024 - 1, p_sfile ) != NULL && *psz_sfile_data != 0)
            {
                p_curr = (srpos_list_t*)malloc( sizeof( srpos_list_t ) );
                if ( p_curr != NULL )
                {
                    p_curr->b_used = false;
                    p_curr->p_next = NULL;
                    p_curr->f_position = atof( psz_sfile_data );
                    if( p_curr->f_position < 0.0 || p_curr->f_position > 1.0 )
                    {
                        p_curr->b_used = true;
                        p_curr->f_position = 0.0;
                    }
                    psz_sfile_uri = strchr( psz_sfile_data, ' ' );
                    while ( psz_sfile_uri && *psz_sfile_uri && psz_sfile_uri[ strlen( psz_sfile_uri ) - 1 ] == '\n' )
                        psz_sfile_uri[ strlen( psz_sfile_uri ) - 1 ] = 0;
                    p_curr->psz_uri = (char *)malloc( strlen( psz_sfile_uri + 1 ) + 1 );
                    if( p_curr->psz_uri == NULL )
                    {
                        free( p_curr );
                        p_curr = NULL;
                    }
                    else
                    {
                        strcpy( p_curr->psz_uri, psz_sfile_uri + 1 );
                        if( p_prev == NULL )
                            p_intf->p_sys->p_list = p_curr;
                        else
                            p_prev->p_next = p_curr;
                        p_prev = p_curr;
                    }
                }
            }
            free( psz_sfile_data );
        }
        fclose( p_sfile );
    }

    var_AddCallback( pl_Get( p_intf ), "item-change", ItemChange, p_intf );

    return VLC_SUCCESS;
}

/*****************************************************************************
 * Close: destroy interface
 *****************************************************************************/
static void Close( vlc_object_t *p_this )
{
    intf_thread_t *p_intf = (intf_thread_t *)p_this;
    srpos_list_t *p_curr = p_intf->p_sys->p_list;
    srpos_list_t *p_next = NULL;
    playlist_t *p_playlist = pl_Get( p_this ); 
    char *psz_sfile_data = NULL;
    FILE *p_sfile = NULL;

    var_DelCallback( p_playlist, "item-change", ItemChange, p_intf );

    /* Save files position info */
    p_sfile = GetSettingsFile( false );
    if( p_sfile )
    {
        psz_sfile_data = (char *)malloc( 32 );
        if( psz_sfile_data )
        {
            while( p_curr != NULL )
            {
                sprintf( psz_sfile_data, "%f", p_curr->f_position );
                fwrite( psz_sfile_data, strlen( psz_sfile_data ), 1, p_sfile );
                fwrite( " ", 1, 1, p_sfile );
                fwrite( p_curr->psz_uri, strlen(p_curr->psz_uri), 1, p_sfile );
                fwrite( "\n", 1, 1, p_sfile );
                p_curr = p_curr->p_next;
            }
            free( psz_sfile_data );
        }
        fclose( p_sfile );
    }

    /* Free memory */
    p_curr = p_intf->p_sys->p_list;
    while( p_curr != NULL )
    {
        free( p_curr->psz_uri );
        p_next = p_curr->p_next;
        free( p_curr );
        p_curr = p_next;
    }
    free( p_intf->p_sys );
}

/*****************************************************************************
 * ItemChange: Playlist item change callback
 *****************************************************************************/
static int ItemChange( vlc_object_t *p_this, const char *psz_var,
                       vlc_value_t oldval, vlc_value_t newval, void *param )
{
    VLC_UNUSED(psz_var);
    VLC_UNUSED(oldval);
    VLC_UNUSED(newval);

    double f_pos = 0.0;
    srpos_list_t *p_file_pos = NULL;
    input_item_t *p_item = NULL;
    intf_thread_t *p_intf = (intf_thread_t *)param;

    playlist_Unlock( (playlist_t *)p_this );    /* playlist_CurrentInput hangs sometimes */
    input_thread_t *p_input = playlist_CurrentInput( (playlist_t *)p_this );

    if( !p_input )
        return VLC_SUCCESS;
    if( p_input->b_dead )
    {
        vlc_object_release( p_input );
        return VLC_SUCCESS;
    }
    p_item = input_GetItem( p_input );
    if( !p_item || !p_item->psz_uri || !*p_item->psz_uri )
    {
        vlc_object_release( p_input );
        return VLC_SUCCESS;
    }

    /* Search for position info */
    p_file_pos = GetPositionInfo( p_intf->p_sys, p_item->psz_uri );

    if( p_file_pos )
    {
        /* Restore last saved position position */
        input_Control( p_input, INPUT_SET_POSITION, p_file_pos->f_position );
    }
    else if( VLC_SUCCESS == input_Control( p_input, INPUT_GET_POSITION, &f_pos) )
    {
        /* Save position info */
        SetPositionInfo( p_intf->p_sys, p_item->psz_uri, f_pos );
    }

    vlc_object_release( p_input );

    return VLC_SUCCESS;
}

/*****************************************************************************
 * GetSettingsFile: Get module settings file
 *****************************************************************************/
static FILE* GetSettingsFile(const bool b_read_only)
{
    FILE *p_file = NULL;
    const char *psz_file_name = MODULE_STRING ".ini";
    char *psz_full_path;
    char *psz_dir;

    psz_dir = config_GetUserDir( VLC_CONFIG_DIR );
    if( !psz_dir )
        return NULL;
    psz_full_path = (char *)malloc( strlen( psz_dir ) + 2 + strlen( psz_file_name ) + 1 );
    if( !psz_full_path )
        return NULL;

    strcpy( psz_full_path, psz_dir );
    strcat( psz_full_path, "/" );
    strcat( psz_full_path, psz_file_name );

    p_file = fopen( psz_full_path, b_read_only ? "r" : "w" );
    
    free( psz_full_path );

    return p_file;
}

/*****************************************************************************
 * GetPositionInfo: Get position info for specified file uri
 *****************************************************************************/
static srpos_list_t* GetPositionInfo( intf_sys_t *p_this, const char *p_uri )
{
    srpos_list_t *p_find = p_this->p_list;
    while( p_find != NULL )
    {
        if( strcmp(p_find->psz_uri, p_uri ) == 0 )
        {
            if( !p_find->b_used )
            {
                /* Reset state */
                p_find->b_used = true;
                return p_find;
            }
            break;
        }
        p_find = p_find->p_next;
    }
    return NULL;
}

/*****************************************************************************
 * SetPositionInfo: Set position info for specified file uri
 *****************************************************************************/
static void SetPositionInfo( intf_sys_t *p_this, const char *p_uri, double f_pos )
{
    srpos_list_t *p_find = p_this->p_list;
    srpos_list_t *p_curr = NULL;
    srpos_list_t *p_prev = NULL;
    int n_max_items = 100;

    /* Search for current record */
    while( p_find != NULL )
    {
        if( strcmp( p_find->psz_uri, p_uri ) == 0 )
        {
            p_curr = p_find;
            break;
        }
        p_prev = p_find;
        p_find = p_find->p_next;
    }

    if( p_curr != NULL )
    {
        if( p_prev != NULL )
            p_prev->p_next = p_curr->p_next;
        else
            p_this->p_list = p_curr->p_next;
        free( p_curr->psz_uri );
        free( p_curr );
    }

    /* Save new record data at the top of list */
    p_curr = (srpos_list_t*)malloc( sizeof( srpos_list_t ) );
    if ( p_curr != NULL )
    {
        p_curr->psz_uri = (char *)malloc( strlen( p_uri ) + 1 );
        if( p_curr->psz_uri == NULL )
        {
            free( p_curr );
            return;
        }
        strcpy( p_curr->psz_uri, p_uri );
        p_curr->f_position = f_pos;
        p_curr->b_used = true;
        p_curr->p_next = p_this->p_list;
        p_this->p_list = p_curr;
    }

    //Remove last records and mark current items
    p_prev = p_this->p_list;
    p_curr = p_this->p_list->p_next;
    while( --n_max_items && p_curr != NULL )
    {
        p_curr->b_used = false;
        p_prev = p_curr;
        p_curr = p_curr->p_next;
    }
    if ( p_curr != NULL)
        p_prev->p_next = NULL;
    while( p_curr != NULL )
    {
        free( p_curr->psz_uri );
        p_find = p_curr->p_next;
        free( p_curr );
        p_curr = p_find;
    }
}
